presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java

```diff
@@ -343,7 +343,7 @@ private PlanWithProperties buildJoinNodeWithPreferredHashes(
 
             return new PlanWithProperties(
                     new JoinNode(
-                            idAllocator.getNextId(),
+                            node.getId(),
```

Comment:
It is definitely not a problem. Everywhere else through the code we reuse the id's for Join node. It might be a problem if there are 2 different nodes with the same id. But here we clearly do the replace.